alerter/src/main/java/org/apache/hertzbeat/alert/calculate/CollectorAlarmHandler.java

```diff
@@ -0,0 +1,199 @@
+                                 AlerterWorkerPool workerPool) {
+        this.offlineAlertMap = new ConcurrentHashMap<>(16);
+        this.alertService = alertService;
+        this.alertCollectorDao = alertCollectorDao;
+        this.workerPool = workerPool;
+        this.bundle = ResourceBundleUtil.getBundle("alerter");
+        List<Collector> collectors = this.alertCollectorDao.findCollectorsByStatus(CommonConstants.COLLECTOR_STATUS_OFFLINE);
+        if (!CollectionUtils.isEmpty(collectors)) {
+            for (Collector collector : collectors) {
+                Map<String, String> tags = new HashMap<>(8);
+                tags.put(CommonConstants.TAG_COLLECTOR_ID, String.valueOf(collector.getId()));
+                tags.put(CommonConstants.TAG_COLLECTOR_NAME, collector.getName());
+                this.offlineAlertMap.put(collector.getName(),
+                        Alert.builder().tags(tags).target(CommonConstants.AVAILABILITY).status(ALERT_STATUS_CODE_PENDING).build());
+            }
+        }
+    }
+
+    /**
+     * handle collector online
+     *
+     * @param identity collector name
+     */
+    public void online(final String identity) {
+        Collector collector = alertCollectorDao.findCollectorByName(identity);
+        if (collector == null) {
+            return;
+        }
+        long currentTimeMill = System.currentTimeMillis();
+        Alert preAlert = offlineAlertMap.remove(identity);
+        if (preAlert != null) {
+            Map<String, String> tags = preAlert.getTags();
+            tags.put(CommonConstants.TAG_COLLECTOR_HOST, collector.getIp());
+            String content = this.bundle.getString("alerter.availability.collector.recover");
+            Alert resumeAlert = Alert.builder()
+                    .tags(tags)
+                    .target(CommonConstants.AVAILABILITY_COLLECTOR)
+                    .content(content)
+                    .priority(CommonConstants.ALERT_PRIORITY_CODE_WARNING)
+                    .status(CommonConstants.ALERT_STATUS_CODE_RESTORED)
+                    .firstAlarmTime(currentTimeMill)
+                    .lastAlarmTime(preAlert.getLastAlarmTime())
+                    .build();
+            workerPool.executeJob(() -> recoverAlarm(identity, resumeAlert));
+        }
+    }
+
+    private void recoverAlarm(String identity, Alert restoreAlert) {
+        List<Long> alertIds = queryAvailabilityAlerts(identity, restoreAlert)
+                .stream()
+                .filter(alert -> Objects.equals(alert.getTags().get(CommonConstants.TAG_COLLECTOR_NAME), identity))
+                .map(Alert::getId)
+                .toList();
+
+        if (!alertIds.isEmpty()) {
+            alertService.editAlertStatus(CommonConstants.ALERT_STATUS_CODE_SOLVED, alertIds);
+
+            // Recovery notifications are generated only after an alarm has occurred
+            alertService.addAlert(restoreAlert);
```

Comment:
👍 hi, how about use `alarmCommonReduce.reduceAndSendAlarm(restoreAlert)`  here? We send it for subsequent alarm convergence, alarm silencing, and alarm storage processing. But it needs to be tested, because before it was monitoring related alarms, I don’t know whether it is compatible with collector alarm.

Reply:
Good idea, I will try to reuse AlarmCommonReduce